src/NuGet.Core/NuGet.Packaging/PackageCreation/Authoring/PackageBuilder.cs

```diff
@@ -649,6 +653,69 @@ private void ValidateIconFile(IEnumerable<IPackageFile> files, string iconPath)
             }
         }
 
+        /// <summary>
+        /// Validate that the readme file is of the correct size/type and can be opened properly.
+        /// </summary>
+        /// <param name="files">Files resolved from the file entries in the nuspec</param>
+        /// <param name="readmePath">readmepath found in the .nuspec</param>
+        /// <exception cref="PackagingException">When a validation rule is not met</exception>
+        private void ValidateReadmeFile(IEnumerable<IPackageFile> files, string readmePath)
+        {
+            if (!string.IsNullOrEmpty(readmePath))
+            {
+                // Validate readme extension
+                var extension = Path.GetExtension(readmePath);
+
+                if (!string.IsNullOrEmpty(extension) &&
+                    !extension.Equals(NuGetConstants.ReadmeExtension, StringComparison.OrdinalIgnoreCase))
+                {
+                    throw new PackagingException(
+                        NuGetLogCode.NU5038, 
+                        string.Format(CultureInfo.CurrentCulture, NuGetResources.ReadmeFileExtensionIsInvalid, readmePath));
+                }
+
+                // Validate entry
+                var readmePathStripped = PathUtility.StripLeadingDirectorySeparators(readmePath);
+
+                var readmeFileList = files.Where(f =>
+                        readmePathStripped.Equals(
+                            PathUtility.StripLeadingDirectorySeparators(f.Path),
+                            PathUtility.GetStringComparisonBasedOnOS()));
```

Comment:
Hmm. What happens in the following scenario:

```xml
<readme>foo\read.md</readme>
```
If I'm working on a cross-platform application, and I'm building on Linux (which uses `/` separators).

Conversely, what if I'm just a Windows user who's used to using `/` as a separator when I type, and I put `<readme>foo/bar.md</readme>` and then try to pack _on Windows_?

Reply:
Good catch! I checked and this was a problem previously, specifically when a user used wildcard characters (eg. '/*'). I wrote tests to catch this out, and also wrote the fix for it. These tests are all passing, even in the CI build (which I think runs them on all platforms?), so this works now!